base/strings/basic.jl

```diff
@@ -606,19 +606,24 @@ julia> first("∀ϵ≠0: ϵ²>0", 1)
 
 julia> first("∀ϵ≠0: ϵ²>0", 3)
 "∀ϵ≠"
+
+julia> first("1234", 10)
+"1234"
 ```
 """
 function first(str::AbstractString, nchar::Integer)
+    s = start(str)
     if 0 <= nchar <= 1
-        return str[1:nchar]
+        return str[s:(s-1+nchar)]
     end
-    str[1:nextind(str, 1, nchar-1)]
+    idx = min(endof(str), nextind(str, s, nchar-1))
+    str[s:idx]
```

Comment:
I guess you can use `@inbounds` here now? That would (or could, if `@boundscheck` is implemented where needed) offset the cost of callind `endof`.

Reply:
I am not sure that `@inbounds` will help (but please correct me if I am wrong or I have misunderstood something what you have meant), because neither
`getindex(s::String, r::UnitRange{Int})` nor `getindex(s::AbstractString, r::UnitRange{Int})`
call `endof` (they work differently, but perform the checks without calling `endof` and what they do does not seem to me to be influenced by `@inbounds` (but maybe I am wrong).

The optimization would be to write a specialized method for `first` for `String` only which would call `unsafe_string` instead of `getindex`. It would be a bit faster, but I do not know if we want to have such a messy code (the same could be done for `last`).

Any thoughts on this?

Reply:
OK. I guess leave it that way for now if the performance impact isn't too large.